### PR TITLE
adds map

### LINKED ISSUE #278: Add map to main site
Add map to main site which will show location of all the user of RDS using [Mapbox GL JS](https://docs.mapbox.com/mapbox-gl-js/guides/)

image

### PR DESCRIPTION
**Pull request by poonammohata**
### Issue
Closes #278 

### What is the change?
Added map, which shows RDS user's location, showing user image as a marker

### Is Development Tested?
- As of right now from the user's API we are not getting location details so for now, tested this with sample data, once the backend changes are complete, need to integrate and test with real data

### Short video of how this will look on the website
https://user-images.githubusercontent.com/54801934/189804189-ccc2f0fb-dce1-482d-8c6b-59b1e9c4a607.mov

### To do while deploying this code
- Need to add this in the build command
```cd js && echo -e "const MAPBOXGL_ACCESS_TOKEN = 'token';\n\nexport default MAPBOXGL_ACCESS_TOKEN;" > config.js```
- The above command will create a config.js file in the js folder, where the token will be defined and exported, which we have imported in the map.js file.
- In the above command in place of a token we need to add the token which we will get on Mapbox after creating a Mapbox account

**Note:**
- Right now this feature is flag based, need to pass `map=true` as query params in the URL, to show the map on the website.
- Markers on the map will be shown for only those users who have both location coordinates and image data.
- If in case no user has location coordinates and image data then- the map wouldn't be shown.

### Things need to be done, for this feature
- [ ] Once the backend changes are complete and we get location data, need to integrate with API and test it thoroughly
- [ ] Need to create an account on [Mapbox GL JS](https://docs.mapbox.com/mapbox-gl-js/guides/) for RDS
- [ ] Need to decide how much zoom will keep on the map, and what will be the center coordinates of the map, right now in the center coordinates I have given coordinates of India, and the zoom is 2.5

### Note to Reviewers: 
- the Percy generate snapshots test will fail as I have made UI changes. That cannot be fixed currently.

### CODE REVIEW

File: css/main.css [hunk 1]
```diff
@@ -695,6 +695,48 @@ footer {
...
+  text-align: center;
+  font-weight: 700;
+  font-size: 2rem;
+  line-height: 150%;
+}
+
+.map-box {
+  height: 300px;
+  overflow: hidden;
+  border: 3px solid #e20665;
```
Comment (rohan09-raj) [inline css/main.css:710]: Can we please use CSS variables here?

File: css/main.css [hunk 2]
```diff
@@ -695,6 +695,48 @@ footer {
...
+#map {
+  overflow: visible;
+  top: 0;
+  bottom: 0;
+  width: 100%;
+  height: 300px;
+}
+
+.mapboxgl-canvas {
+  width: 100% !important;
```
Comment (rohan09-raj) [inline css/main.css:723]: Is there any other to achieve the same without using `!important`?

File: js/map.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      );
+      addMarkersToMap(userHaveLocationAndImageDetails);
+    }
+  } catch (err) {
+    console.log('get users data error', err);
+  }
+};
+
+async function addMarkersToMap(userDetails) {
+  if (map._loaded && map._fullyLoaded && userDetails.length && showMap) {
```
Comment (rohan09-raj) [inline js/map.js:43]: Can we simplify these conditions by using reverse conditions means conditions with ! (not) operator?

File: js/map.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    userDetails.forEach(({ location_coordinates, picture }) => {
+      // Create a DOM element for each marker.
```
Comment (rohan09-raj) [inline js/map.js:45]: Line 46 is self explanatory, no need for this comment

File: js/map.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      const el = document.createElement('div');
```
Comment (rohan09-raj) [inline js/map.js:46]: Can we name it better please ?

File: js/map.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      el.className = 'marker';
+      el.style.backgroundImage = `url(${picture.url})`;
+      // Add markers to the map.
+      new mapboxgl.Marker(el).setLngLat(location_coordinates).addTo(map);
+      mapSection.classList.remove('element-display-remove');
+      map.resize();
+    });
+  } else {
+    return;
+  }
```
Comment (rohan09-raj) [inline js/map.js:56]: We can simplify this if else by using guard condition. Something like below
```
if (conditions) {
  return;
}

// rest of the code, here no need to put the if conditions
```

File: css/main.css (hunk 2 above)
Comment (poonammohata) [reply css/main.css:723]: In the Mapbox library, in this canvas element, they have used inline CSS for width, so other than using !important, I didn't find anything to overwrite that style.

image

If you know, of any other way by which we can change the width, then please let me know.

File: css/main.css [hunk 7]
```diff
@@ -695,6 +695,48 @@ footer {
...
+  border: 3px solid var(--color-events);
+  margin: 2% 10%;
+}
+
+#map {
+  overflow: visible;
+  top: 0;
+  bottom: 0;
+  width: 100%;
+  height: 300px;
```
Comment (rohan09-raj) [inline css/main.css:719]: Can we please make the units consistent here?

File: css/main.css [hunk 8]
```diff
@@ -695,6 +695,48 @@ footer {
...
+  font-weight: 700;
+  font-size: 2rem;
+  line-height: 150%;
+}
+
+.map-box {
+  height: 300px;
+  overflow: hidden;
+  border: 3px solid var(--color-events);
+  margin: 2% 10%;
```
Comment (rohan09-raj) [inline css/main.css:711]: Can we please use pixels for margins and paddings?

File: css/main.css (hunk 2 above)
Comment (rohan09-raj) [reply css/main.css:723]: Got it, thanks for explanation, resolving the conversation

Comment (rohan09-raj) [reply css/main.css:723]: Do Mapbox library not provide the CSS customization??
Also, why do you exactly want to change the width? What is the purpose here?

Comment (poonammohata) [reply css/main.css:723]: After updating css, not needed this anymore, removed it

File: css/main.css (hunk 8 above)
Comment (rohan09-raj) [reply css/main.css:711]: Ahh ok, then % is fine I feel

**Review by rohan09-raj — APPROVED** [review]
LGTM 💯 Approving and merging the PR

**Pull request merged**